scrape/scrape.go

```diff
@@ -1533,15 +1537,17 @@ loop:
 			lset = ce.lset
 
 			// Update metadata only if it changed in the current iteration.
-			sl.cache.metaMtx.Lock()
-			metaEntry, metaOk := sl.cache.metadata[yoloString([]byte(lset.Get(labels.MetricName)))]
-			if metaOk && metaEntry.lastIterChange == sl.cache.iter {
-				shouldAppendMetadata = true
-				meta.Type = metaEntry.Type
-				meta.Unit = metaEntry.Unit
-				meta.Help = metaEntry.Help
+			if sl.appendMetadataToWAL {
+				sl.cache.metaMtx.Lock()
+				metaEntry, metaOk := sl.cache.metadata[yoloString([]byte(lset.Get(labels.MetricName)))]
+				if metaOk && metaEntry.lastIterChange == sl.cache.iter {
+					shouldAppendMetadata = true
+					meta.Type = metaEntry.Type
+					meta.Unit = metaEntry.Unit
+					meta.Help = metaEntry.Help
+				}
+				sl.cache.metaMtx.Unlock()
```

Comment:
since this and the later block on lines 1576 to 1585 appear to be exactly the same can we move it to a single anonymous function?

Reply:
These blocks have a subtle difference. For series that we've already seen, we're checking if the metadata changed in the current iteration, but for new series we're only checking whether metadata exists at all.

I've extracted the logic of both places into a shared anonymous function as you proposed in https://github.com/prometheus/prometheus/pull/10312/commits/ce5b1f0263e84f99150deb9e0c321d2bd73b7212. Let me know how it looks to you, hopefully it simplifies instead of complicating things!